_pytest/doctest.py

```diff
@@ -75,14 +78,72 @@ def _is_doctest(config, path, parent):
 
 class ReprFailDoctest(TerminalRepr):
 
-    def __init__(self, reprlocation, lines):
-        self.reprlocation = reprlocation
-        self.lines = lines
+    def __init__(self, reprlocation_lines):
+        # List of (reprlocation, lines) tuples
+        self.reprlocation_lines = reprlocation_lines
 
     def toterminal(self, tw):
-        for line in self.lines:
-            tw.line(line)
-        self.reprlocation.toterminal(tw)
+        for reprlocation, lines in self.reprlocation_lines:
+            for line in lines:
+                tw.line(line)
+            reprlocation.toterminal(tw)
+
+
+class DoctestFailureContainer(object):
+
+    NAME = 'DocTestFailure'
+
+    def __init__(self, test, example, got):
+        self.test = test
+        self.example = example
+        self.got = got
+
+
+class DoctestUnexpectedExceptionContainer(object):
+
+    NAME = 'DoctestUnexpectedException'
+
+    def __init__(self, test, example, exc_info):
+        self.test = test
+        self.example = example
+        self.exc_info = exc_info
+
+
+def _init_runner_class():
+    import doctest
+
+    class PytestDoctestRunner(doctest.DocTestRunner):
+        """
+        Runner to collect failures.  Note that the out variable in this case is
+        a list instead of a stdout-like object
+        """
+        def report_start(self, out, test, example):
+            pass
+
+        def report_success(self, out, test, example, got):
+            pass
+
+        def report_failure(self, out, test, example, got):
+            out.append(DoctestFailureContainer(test, example, got))
+
+        def report_unexpected_exception(self, out, test, example, exc_info):
+            out.append(DoctestUnexpectedExceptionContainer(test, example, exc_info))
+
+    return PytestDoctestRunner
+
+
+def _get_runner(*args, **kwargs):
+    # We need this in order to do a lazy import on doctest
+    global RUNNER_CLASS
+    if RUNNER_CLASS is None:
+        RUNNER_CLASS = _init_runner_class()
+    return RUNNER_CLASS(*args, **kwargs)
+
+
+class DoctestFailure(Exception):
```

Comment:
I suggest we call this something which does not sound so similar to `doctest.DocTestFailure` to avoid confusion, how about `DoctestMultipleFailures`?